design/one-pager-cross-resource-referencing.md

```diff
@@ -0,0 +1,296 @@
+
+* **Composite Resource Attribute**
+
+  We define the attribute `foo` of resource `r` *composite*, if `foo`'s value is
+  deterministic and composed of the attributes of other resources. As an
+  example, `network` attribute of `Subnetwork` type in GCP, is formed as
+
+  `/projects/[gcp-project]/global/networks/[network-name]`
+
+  where `[gcp-project]` and `[network-name]` are attributes of other resources.
+
+## Objectives
+
+### Goals
+
+* Support `gitops` style declarative resource provisioning. Apply a directory of
+  YAML resources, which will eventually become a series of online and
+  functioning managed resources. This requires:
+  * Support to reference attributes of other resources in a resource YAML object
+  * Support for cross resource blocking dependency
+
+* Support using existing external resources to be referenced by the YAML objects
+
+### Non Goals
+
+* Support non-blocking dependencies between resources
+
+## Proposal
+
+Let's consider the following sample solution where a `VPC`, and a `Subnet` needs to be
+provisioned. The YAML object will look like following:
+
+```yaml
+---
+apiVersion: network.aws.crossplane.io/v1alpha2
+kind: VPC
+metadata:
+  namespace: cool-ns
+  name: my-vpc
+spec:
+  ...
+---
+apiVersion: network.aws.crossplane.io/v1alpha2
+kind: Subnet
+metadata:
+  namespace: cool-ns
+  name: my-subnet
+spec:
+  # this is the vpcId of the external vpc, represented by my-vpc
+  vpcId: [my-vpc_vpcId]
+  ...
+```
+
+### Cross referencing using Reference Resolvers
+
+In this example since the `vpcId` is non-deterministic, `[my-vpc_vpcId]` is not
+known prior to provisioning the `VPC`. Therefore we will need a mechanism to
+indicate this cross reference in the YAML object. To solve this, I propose the
+notion of *Reference Resolver*, as a `go` interface as following:
+
+```go
+type ReferenceResolver interface {
+  // resolves the attribute, or returns an error
+  Resolve() (string,error)
+}
+```
+
+Having this interface, we can implement `VpcIDRefResolver` as
+
+```go
+type VpcIDRefResolver struct {
+  // the list of objects that are needed for resolving vpcID
+  // in this case we only need VPC name, but we could reference more objects
+  VPCName string `json:"vpcName"`
+}
+
+func (r *VpcIDRefResolver) Resolve()(string,error){
+  // implementation for building the attribute (in this case vpcID)
+  // from the referenced objects( in this case a VPC)
+}
+```
+
+Then we can modify [Subnet type] as:
+
+```diff
+- // VPCID is the ID of the VPC.
+- VPCID string `json:"vpcId"`
++ // VPCIDRef resolves the VPCID from the refenreced VPC
++ VPCIDRef VpcIDRefResolver `json:"vpcIdRef"`
+```
+
+Now we can update the `Subnet` YAML object in the sample solution as following:
+
+```yaml
+---
+apiVersion: network.aws.crossplane.io/v1alpha2
+kind: Subnet
+metadata:
+  namespace: cool-ns
+  name: my-subnet
+spec:
+  # reference to API objects from which the vpcId will be resolved
+  vpcIdRef:
```

Comment:
I am seeing that my comment seems to have some overlap with Bassam's earlier comment and Nic's response, though I am still wondering about the idea of specifying references in a way that feels more like the way it is done for pod value injection.

Reply:
Hi @suskin, thanks for the input!

#707 initial proposal was something similar, even though I wasn't thinking of this nice `fieldRef` syntax  of `Pod` env variables that you are pointing here. As I have tried to demonstrate in the proposal, the referencing itself is not the sole challenge here. We also want to resolve it (e.g. wait for a resource to be created and then use it)

So even though I agree that having an already established pattern is preferred, I don't think this pattern will address our problem on its own without any modification.
The other related work is Google Config Connector, where they are using similar approach as what is proposed here:
https://cloud.google.com/config-connector/docs/how-to/creating-resource-references


The other goal of this design is to hard-code as much logic as possible into the types and their controllers, rather than manifesting it in the YAML object. For instance,  instead of 
```
spec:
  vpcId:
    objRef: my-vpc
      fieldPath: status.id
```

we will have 

```
spec:
  vpcIdRef:
    vpcName: my-vpc
```
And the logic (e.g. the resolver) itself takes care of `fieldPath: status.id` part.

Reply:
I believe we're contrasting two options here - more opinionated and thus more succinct, vs more flexible and thus more verbose.

Currently this design proposes the more opinionated and more succinct path:

```yaml
spec:
  vpcRef:
    name: my-vpc
```

If we only reference a predetermined `kind` - i.e. in the above case `kind: VPC` - I don't think there's any value in including a `fieldPath`. If you're referring to a `kind: VPC` I can't imagine any scenario in which the fieldPath would be set to anything but `.status.id`. If there is only ever one valid field, why expose that as a setting to users?

Where I think the `fieldPath` pattern becomes more compelling is if you allow a reference to an object of an arbitrary kind. From a user's perspective this would allow you to configure an intent like "put this EKS cluster in the VPC used by that RDS instance". However, this comes at the expense of a much more complicated API:

```yaml
spec:
  vpcRef:
    apiVersion: network.aws.crossplane.io/v1alpha2
    kind: VPC
    name: my-vpc
    fieldPath: status.id
```

or...
```yaml
spec:
  vpcRef:
    apiVersion: compute.aws.crossplane.io/v1alpha2
    kind: EKSCluster
    name: my-eks
    fieldPath: spec.vpcID
```

My feeling is that the latter API's additional value does not justify its additional complexity and cognitive burden. The first API still allows you to configure that an RDS instance use the same VPC as an EKS cluster - it just requires that you either model that VPC in Crossplane to do so, or use a plain string reference rather than a resource reference.

It's also worth noting that the succinct, opinionated path is a strict subset of the flexible, verbose path, so we could always start opinionated and add flexibility without any breaking API changes if we get feedback that our initial simpler design is too restrictive.